### DISCUSSION TITLE
Add a way to force search results to be exhaustive

### DISCUSSION THREAD

**Discussion by gmourier** [meilisearch/product#335] [Feedback & Feature Proposal]
## Description

We are opening this discussion to gather and consolidate feedback that are related to the fact that it is not possible to force MeiliSearch to return an exhaustive number of results.

The initial idea would be to have a way to force the search to return exhaustive results knowing that it will impact performance in a negative way.

### Related discussions

- https://github.com/meilisearch/product/discussions/327
- https://github.com/meilisearch/MeiliSearch/issues/1993
- https://github.com/meilisearch/MeiliSearch/discussions/1929

**Comment by homerjam** [comment]
I would like this feature as I'm using Meilisearch as the primary method of querying/filtering/sorting data in my stack, a small performance penalty would be an acceptable compromise and I will implement a cache in any case.

Another reason I would like this feature is in order to retrieve a definitive list of all the IDs in an index for verification that everything has synced from the primary database. Alternatively, a simple method just to get those IDs would work.

**Reply by gmourier** [reply]
Hello @homerjam 

> Another reason I would like this feature is in order to retrieve a definitive list of all the IDs in an index for verification that everything has synced from the primary database. Alternatively, a simple method just to get those IDs would work.

Did you tried to use the `indexes/{indexUid}/documents` endpoint in combination with `limit`/ `offset` parameters? You can also only retrieve the `id` by providing `attributesToRetrieve=id` as a query parameter

**Comment by curquiza** [comment]
More explanation about why the `nbHits` returned by Meilisearch is not exhaustive: https://github.com/meilisearch/product/discussions/467

**Comment by ManyTheFish** [comment]
Hello @homerjam!
We are working on improving pagination in Meilisearch and we just released a first prototype that makes creating numbered pagination way easier! :fire:
It would be super helpful if you could let us know if this helps with your issue.
To know more about the prototype and how to use it, take a look at [the dedicated discussion](https://github.com/meilisearch/meilisearch/discussions/2635).

**Comment by mmachatschek** [comment]
on our end, the `estimatedTotalHits` returned a fairly accurate number until we started to use the distinctAttribute feature https://docs.meilisearch.com/reference/api/settings.html#distinct-attribute

now the numbers are off by ~20%

having the option to tell the search that we want to receive an exact count would be great.

possible fixes to work around this issue:

- add a specific attribute to one of the documents that is distinguished by the distinctAttribute
- fetch all documents in the index for this search (limit: 1000000) and get the exact count

**Reply by ManyTheFish** [reply]
Hey @mmachatschek,
this feature is included in the current [pre-release of V0.30](https://github.com/meilisearch/meilisearch/releases/tag/v0.30.0rc0),
if you want more information, your issue seems similar to the below discussion and could help you to solve it:
https://github.com/meilisearch/product/discussions/483#discussioncomment-4017732

Thank you for the interest you have in our product!

**Reply by mmachatschek** [reply]
Hi @ManyTheFish,

thanks for your reply. The changelog is not really explicit about this.

so if I use the new pagination workflow (page, hitsPerPage), I will get a `totalHits` count that is like an `exhaustiveTotalHits`.

the `estimatedTotalHits` is for a `limit, offset` search.

I need to try this out, if this works with the `distinctAttribute` setting

**Reply by curquiza** [reply]
Hello @mmachatschek 😄 

how could we make the changelog clearer? Should we mention we don't need to use `exhaustiveTotalHits`?

**Reply by mmachatschek** [reply]
hi @curquiza after reading it three times now it starts to make sense. probably I'm just not reading careful enough.

> These new search parameters can be used in the place of offset and limit when creating interfaces that require knowing the exhaustive number of search results.

this mentions it.

**Reply by curquiza** [reply]
Thanks @mmachatschek, we have a PR regarding the changelog, we will try to update this, you can follow the discussion (and participate if you want and have the time) [here](https://github.com/meilisearch/core-team/pull/34/files#r1012763086)

**Reply by mmachatschek** [reply]
hi there @ManyTheFish  @curquiza 

I wasn't able to test v0.30 until now but for me, using `page` and `hitsPerPage` do not show an exhaustive number of hits.

using different values of page and hits per page show different numbers.

I can share the dataset privately on discord if needed

Check this (the index data is not changing while testing):

(page: 0, hitsPerPage 10, results: 2539)
![image](https://user-images.githubusercontent.com/10237069/205693917-077e0e5a-e054-4314-a3b0-0973379be287.png)

(page: 1, hitsPerPage 10, results: 2531)
![image](https://user-images.githubusercontent.com/10237069/205694198-7688101e-5cbe-44dc-968c-adfa41a06c3b.png)

(page 1, hitsPerPage 20, results: 2510)
![image](https://user-images.githubusercontent.com/10237069/205694418-71760222-0d69-4896-9f08-468631891606.png)

(page 2, hitsPerPage 10, results: 2510)
![image](https://user-images.githubusercontent.com/10237069/205694564-545d03d9-d3c8-464b-bc13-f9a2716959cc.png)

(page 2, hits per Page 20, results: 2479)
![image](https://user-images.githubusercontent.com/10237069/205694734-ab73d71d-6394-4871-887a-de703717cdf1.png)

**Reply by gmourier** [reply]
Thank you @mmachatschek !

**Reply by mmachatschek** [reply]
I created an issue from my comment here: https://github.com/meilisearch/meilisearch/issues/3200

**Comment by gmourier** [comment]
Hey everyone 👋

[`v0.30`](https://github.com/meilisearch/meilisearch/releases/tag/v0.30.1) introduced a new way to navigate search results. It's now possible to implement an exhaustive page selection component. 🪄 

📚 More information [here](https://docs.meilisearch.com/learn/advanced/pagination.html#numbered-page-selectors).

**Comment by gmourier** [comment]
Hey everyone 👋

I'm locking that discussion.

Please open a new discussion if you want to share feedback / ask for a feature request.

Thank you! 🙇‍♂️
